pci/src/vfio.rs

```diff
@@ -1885,3 +1898,88 @@ impl Snapshottable for VfioPciDevice {
 }
 impl Transportable for VfioPciDevice {}
 impl Migratable for VfioPciDevice {}
+
+/// This structure implements the ExternalDmaMapping trait. It is meant to
+/// be used when the caller tries to provide a way to update the mappings
+/// associated with a specific VFIO container.
+pub struct VfioDmaMapping<M: GuestAddressSpace> {
+    container: Arc<VfioContainer>,
+    memory: Arc<M>,
+    mmio_regions: Arc<Mutex<Vec<MmioRegion>>>,
+}
+
+impl<M: GuestAddressSpace> VfioDmaMapping<M> {
+    /// Create a DmaMapping object.
+    /// # Parameters
+    /// * `container`: VFIO container object.
+    /// * `memory`: guest memory to mmap.
+    /// * `mmio_regions`: mmio_regions to mmap.
+    pub fn new(
+        container: Arc<VfioContainer>,
+        memory: Arc<M>,
+        mmio_regions: Arc<Mutex<Vec<MmioRegion>>>,
+    ) -> Self {
+        VfioDmaMapping {
+            container,
+            memory,
+            mmio_regions,
+        }
+    }
+}
+
+impl<M: GuestAddressSpace + Sync + Send> ExternalDmaMapping for VfioDmaMapping<M> {
+    fn map(&self, iova: u64, gpa: u64, size: u64) -> std::result::Result<(), io::Error> {
+        let mem = self.memory.memory();
+        let guest_addr = GuestAddress(gpa);
+        let user_addr = 'block: {
+            // Check if the address is within RAM
+            if mem.check_range(guest_addr, size as usize) {
+                break 'block mem.get_host_address(guest_addr).unwrap() as u64;
+            }
+
+            // Check if the address is within an mmio region
+            for region in self.mmio_regions.lock().unwrap().iter() {
```

Comment:
I think that this logic would benefit from some simplification. (e.g.)
```rust
if mem.check_range(guest_addr, size as usize) {
    mem.get_host_address(guest_addr).unwrap() as u64;
} else if mmio_regions.check_range(guest_addr, size as usize) {
    mmio_regions.get_host_address(guest_addr).unwrap() as u64;
} else {
   return error
}
```

Reply:
Yep that makes sense, this should be addressed - check_range and get_host_address methods have both been created for Vec<MmioRegions>